third_party/gtp_ovs/ovs-gtp-patches/2.15/0024-QFI-Support-in-OVS.patch

```diff
@@ -0,0 +1,1021 @@
++AT_CHECK([echo 'module openvswitch +p' > /sys/kernel/debug/dynamic_debug/control])
++
++dnl Now add rules with qfi for OVS to forward to the tunnel and local port
++AT_CHECK([ovs-ofctl add-flow br0 "priority=100 ip,tun_id=0,nw_dst=10.1.1.2,qfi=6 action=output:at_gtp0"])
++
++dnl Now add rules for OVS to forward to the tunnel and local port
++AT_CHECK([ovs-ofctl add-flow br0 "priority=1 action=drop"])
++
++sleep 1
++NS_CHECK_EXEC([at_ns0], [python3 $srcdir/gtp-packet.py 172.31.1.1 172.31.1.100 10.1.1.1 10.1.1.2 5555 p0 False 1234 2>/dev/null], [0], [dnl
++.
++Sent 1 packets.
++])
++sleep 1
++
++AT_CHECK([ovs-ofctl dump-flows br0 | ofctl_strip | grep ip], [0], [dnl
++ n_packets=1, n_bytes=43, priority=100,ip,tun_id=0,qfi=6,nw_dst=10.1.1.2 actions=output:1
++])
++dnl sleep 10000
++OVS_TRAFFIC_VSWITCHD_STOP
++AT_CLEANUP
++
++
++AT_SETUP([layer3 - Qfi value set for downlink over GTP])
++OVS_TRAFFIC_VSWITCHD_START([set Bridge br0 other-config:hwaddr="00:12:34:56:78:bb"])
++OVS_CHECK_GTP_L3()
++OVS_CHECK_MIN_KERNEL(4,10)
++
++ADD_BR([br-underlay])
++
++ADD_NAMESPACES(at_ns0)
++
++dnl Set up underlay link from host into the namespace using veth pair.
++ADD_VETH(p0, at_ns0, br-underlay, "172.31.1.1/24")
++AT_CHECK([ip addr add dev br-underlay "172.31.1.100/24"])
++AT_CHECK([ip link set dev br-underlay up])
++AT_CHECK([modprobe vport_gtp])
++
++dnl Set up tunnel endpoints on OVS outside the namespace and with a native
++dnl linux device inside the namespace.
++
++ADD_OVS_TUNNEL([gtpu], [br0], [at_gtp0], [172.31.1.1], [10.1.1.2/24])
++AT_CHECK([ip neigh add 10.1.1.1 lladdr 00:12:34:56:78:aa dev br0])
++NS_CHECK_EXEC([at_ns0], [gtp-link add at_gtp1 --sgsn &], [0], [ignore])
++dnl kernel 4.9
++dnl NS_CHECK_EXEC([at_ns0], [gtp-tunnel add at_gtp1 v1 0 0 10.1.1.2 172.31.1.100], [0], [ignore], [ignore])
++NS_CHECK_EXEC([at_ns0], [gtp-tunnel add at_gtp1 v1 0 0 10.1.1.1 172.31.1.100], [0], [ignore], [ignore])
++NS_CHECK_EXEC([at_ns0], [ip addr add dev at_gtp1 10.1.1.1/24])
++NS_CHECK_EXEC([at_ns0], [ip link set dev at_gtp1 mtu 1450 up])
++NS_CHECK_EXEC([at_ns0], [ip link set dev p0 mtu 1480 up])
++
++AT_CHECK([ovs-ofctl add-flow br-underlay "actions=normal"])
++
++AT_CHECK([ovs-appctl vlog/set dbg], [0], [ignore])
++AT_CHECK([echo 'module openvswitch +p' > /sys/kernel/debug/dynamic_debug/control])
++
++dnl Now add rules with qfi for OVS to forward to the tunnel and local port
++AT_CHECK([ovs-ofctl add-flow br0 "priority=100 ip,nw_dst=10.1.1.2 action=load:0->NXM_NX_TUN_ID[],load:6->NXM_NX_QFI[],output:local"])
++sleep 1
++
++NS_CHECK_EXEC([at_ns0], [python3 $srcdir/gtp-packet.py 172.31.1.1 172.31.1.100 10.1.1.1 10.1.1.2 5555 p0 False 1234 2>/dev/null], [0], [dnl
++.
++Sent 1 packets.
++])
++sleep 1
++
++AT_CHECK([ovs-ofctl -O OpenFlow13 dump-flows | ofctl_strip ], [0], [dnl
++ n_packets=1, n_bytes=43, priority=100,ip,nw_dst=10.1.1.2 actions=load:0->NXM_NX_TUN_ID[],load:0x6->NXM_NX_QFI[],LOCAL
++])
++dnl sleep 10000
++OVS_TRAFFIC_VSWITCHD_STOP
++AT_CLEANUP
++
```

Comment:
Add flow action parsing for QFI field similar to existing gtpu() action.

Reply:
Yes, done.